File: Question 0082 - Remove Duplicates from Sorted List II/src/Solution.java
class Solution {
    public ListNode deleteDuplicates(ListNode head)
    {
       if(head == null) return head;

       ListNode curr = head;
       ListNode next = curr.next;


       ListNode ans = new ListNode(-1);
       ListNode ansPtr = ans;

       while(next != null)
       {
           if(curr.val == next.val)
           {
               while(next != null && curr.val == next.val)
               {
                   curr = next;
                   next = curr.next;
               }

               //move to the next non duplicate val
               curr = next;
               if(next != null) next = curr.next;
           }
           else
           {
               ansPtr.next = curr;
               ansPtr = curr;
               curr = curr.next;
               next = curr.next;
           }
       }

        // To handle the case where the last node is not a duplicate
       ansPtr.next = curr;

       return ans.next;
    }
}

File: Question 0082 - Remove Duplicates from Sorted List II/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        if(head == NULL) return head;
        ListNode ans(-1);
        ListNode *ansPtr = &ans;

        ListNode *curr = head;
        ListNode *next = curr->next;

        while (next != NULL)
        {
            if (curr->val == next->val)
            {
                while (next != NULL && curr->val == next->val)
                {
                    curr = next;
                    next = curr->next;
                }

                curr = next;
                if (next != NULL) next = curr->next;
            }
            else
            {
                ansPtr->next = curr;
                ansPtr = curr;
                curr = curr->next;
                next = curr->next;
            }
        }

        ansPtr->next = curr;
        return ans.next;
    }
};